mlflow/store/artifact/optimized_s3_artifact_repo.py

```diff
@@ -0,0 +1,275 @@
+            extra_args["ContentEncoding"] = guessed_encoding
+        environ_extra_args = self.get_s3_file_upload_extra_args()
+        if environ_extra_args is not None:
+            extra_args.update(environ_extra_args)
+        s3_client.upload_file(Filename=local_file, Bucket=bucket, Key=key, ExtraArgs=extra_args)
+
+    def log_artifact(self, local_file, artifact_path=None):
+        artifact_file_path = os.path.basename(local_file)
+        if artifact_path:
+            artifact_file_path = posixpath.join(artifact_path, artifact_file_path)
+        self._upload_to_cloud(
+            cloud_credential_info=self._get_s3_client(),
+            src_file_path=local_file,
+            artifact_file_path=artifact_file_path,
+        )
+
+    def _get_write_credential_infos(self, remote_file_paths):
+        """
+        Instead of returning ArtifactCredentialInfo objects, we instead return a list of initialized
+        S3 client. We do so because S3 clients cannot be instantiated within each thread.
+        """
+        return [self._get_s3_client() for _ in remote_file_paths]
+
+    def _upload_to_cloud(self, cloud_credential_info, src_file_path, artifact_file_path):
+        dest_path = posixpath.join(self.bucket_path, artifact_file_path)
+        key = posixpath.normpath(dest_path)
+        if (
+            MLFLOW_ENABLE_MULTIPART_UPLOAD.get()
+            and os.path.getsize(src_file_path) > _MULTIPART_UPLOAD_CHUNK_SIZE
+        ):
+            self._multipart_upload(cloud_credential_info, src_file_path, self.bucket, key)
+        else:
+            self._upload_file(cloud_credential_info, src_file_path, self.bucket, key)
+
+    def _multipart_upload(self, cloud_credential_info, local_file, bucket, key):
+        # Create multipart upload
+        s3_client = cloud_credential_info
+        response = s3_client.create_multipart_upload(Bucket=bucket, Key=key)
+        upload_id = response["UploadId"]
+
+        # Create presigned URL for each part
+        num_parts = math.ceil(os.path.getsize(local_file) / _MULTIPART_UPLOAD_CHUNK_SIZE)
+
+        def _get_presigned_upload_part_url(part_number):
+            return s3_client.generate_presigned_url(
+                "upload_part",
+                Params={
+                    "Bucket": bucket,
+                    "Key": key,
+                    "UploadId": upload_id,
+                    "PartNumber": part_number,
+                },
+            )
+
+        presigned_urls = [
+            _get_presigned_upload_part_url(part_number) for part_number in range(1, num_parts + 1)
+        ]
+
+        # define helper functions for uploading data
+        def _upload_part(presigned_url, local_file, start_byte, size):
+            data = read_chunk(local_file, size, start_byte)
+            with cloud_storage_http_request("put", presigned_url, data=data) as response:
+                augmented_raise_for_status(response)
+                return response.headers["ETag"]
```

Comment:
Not a blocker, but should we retry chunk uploads as a follow-up? In the current implementation, no chunk upload failure is allowed.

Reply:
+1

Reply:
In the original `DatabricksArtifactRepository` logic, we retried only on authorization failures. So any non-auth related chunk upload failures would ruin the entire upload. Given that the whole OptimizedS3Repository is initialized on a scoped token, I thought it didn't make sense to retry due to auth failures because a new presigned URL would have the same auth as a previously generated one. Open to suggestions for a follow-up PR.

Reply:
What errors does boto3 retry on?

Reply:
Some 50x error codes. See more details here: https://boto3.amazonaws.com/v1/documentation/api/latest/guide/retries.html

Reply:
Got it. Something we should consider implementing. I'm not sure how stable/unstable chunk uploads/downloads are.